lib/mutant/isolation.rb

```diff
@@ -33,30 +37,41 @@ module Fork
       #   if block terminates abnormal
       #
       # rubocop:disable MethodLength
-      def self.call(&block)
-        IO.pipe(binmode: true) do |reader, writer|
+      def call(&block)
+        io.pipe(binmode: true) do |reader, writer|
           writer.binmode
           begin
-            pid = Process.fork do
-              File.open(File::NULL, File::WRONLY) do |file|
-                $stderr.reopen(file)
-                reader.close
-                writer.write(Marshal.dump(block.call))
-                writer.close
-              end
+            pid = process.fork do
+              child(reader, writer, &block)
             end
 
             writer.close
-            Marshal.load(reader.read)
+            marshal.load(reader.read)
           ensure
-            Process.waitpid(pid) if pid
+            process.waitpid(pid) if pid
           end
         end
       rescue => exception
         raise Error, exception
       end
 
-    end # Fork
+      # Handle child process
+      #
+      # @param [IO] reader
+      # @param [IO] writer
+      #
+      # @return [undefined]
+      def child(reader, writer, &block)
+        reader.close
 
+        devnull.call do |null|
+          stderr.reopen(null)
+          stdout.reopen(null)
+          writer.write(marshal.dump(block.call))
```

Comment:
How could the write be interleaved? We only have one writer and one reader. OT: I also think we should bypass the buffers here, because "why rebuffering" something that is already in a "string buffer ready to be handed over to the kernel".

But from terms for "write atomitcy in this case" `writer.write(buffer); writer.close` has the same smeantics than `writer.syswrite(buffer); writer.close`.